### ISSUE TITLE
TSServer protoco.d.ts refers to types from typescriptServices

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/TypeScript#11482]
**TypeScript Version:**  2.0.6-insider
**Code**

The protocol.d.ts file refers to types from typeScriptServices. Since we pull the protocol.d.ts into VSCode to program against the proper version of the protocol the file should be fully self contained (the file is basically a specification). Types I recognized are missing:
```
TodoCommentDescriptor, EditorSettings, ScriptKind, TypingOptions, CompilerOptions, ts.TextChange,
ts.FormatCodeOptions
```

There is one additional type that is incorrectly referenced `protocol.ProjectVersionInfo`. Should only be `ProjectVersionInfo`

**Comment by dbaeumer** [comment]
@mhegazy @vladima can you please have a look. This makes it very hard to adopt the 2.0.6 branch.

**Comment by mhegazy** [comment]
I do not think we ever mean the protocle.d.ts to be a used in vacuum. It is easier and more maintainable for us to reuse definitions for common types instead of having to duplicate them all. not sure i see why this would be a blocker.

**Comment by vladima** [comment]
We can generate `protocol.d.ts` automatically by transitively pulling all referenced types into final `.d.ts` file. However I don't think it meets the bar to be fixed in `2.0.6` given that workaround exists - just copy all listed declarations from `typescriptServices`.

**Comment by dbaeumer** [comment]
It is a half blocker since it requires manual work on our end. With the work I did yesterday I consume the tsserver 'automatically' in our builds. But this only works nicely if I can consume the protocol.d.ts automatically as well. 

I will handcraft a protocol.d.ts file for now. But please ping me when you guys make changes that are incompatible because it requires changing the protocol.d.ts file on our end.

We do have the same problem with our standalone editor as well and @alexandrudima wrote tooling to generate a single 'monaco.d.ts' file form n other files by selectively copying types. You might want to have a look at his tooling. May be you can borrow some code. 

Alternatively you could think about specify the protocol in JSON schema and then clients can use tools to generate a protocol.d.ts. VS Code's debugger protocol is doing that and I have requests to do the same for the language server protocol.

**Comment by dbaeumer** [comment]
OK. I merged this by hand and it almost took me two hours.

To ensure that the protocol.d.ts is consistent (see the type mismatches I discovered) and doesn't introduce any unwanted breakage I would like to ask you to make this clean for 2.0.6.

**Comment by mhegazy** [comment]
@vladima the enums in the protocol would not work for @dbaeumer, we need to make sure any enum in this file is a `const enum`.

**Comment by vladima** [comment]
ok, will do

**Comment by mhegazy** [comment]
@vladima @dbaeumer wants to use the file as a module.. so let's say he adds `export = ts.server.prototcol;` at the end. now the import for this module will not be elided because the const enums.. any thoughts?

**Comment by vladima** [comment]
what about we keep the content as is but will make it `.ts` instead of `.d.ts` - then const enums will not be necessary

**Comment by mhegazy** [comment]
that is possible, but then ppl will have to build it with their code.. may not be that bad..

**Comment by mhegazy** [comment]
these seems to be another .d.ts emitter bug that @dbaeumer is running into as well, i will look into it.

**Comment by mhegazy** [comment]
alternatively, why do we use enums? can we use union types instead?

**Comment by mhegazy** [comment]
strings are more readable than numbers. it means we have to add parsing for it.. but that might be for the best.

**Comment by dbaeumer** [comment]
Thanks for looking into this. Currently I have a protocol.ts which I compile. However a solution where a compile is not necessary would be better since then I can refer to protocol.d.ts in the node_modules folder.

**Comment by vladima** [comment]
should be addressed by #11651
